router/pkg/pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,380 @@
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for _, topic := range sub.topics {
+		if _, ok := p.subscriptions[topic]; !ok {
+			p.subscriptions[topic] = make([]*subscription, 0, 10)
+		}
+		p.subscriptions[topic] = append(p.subscriptions[topic], sub)
+	}
+}
+
+// topicPoller polls the Kafka topic for new records and calls the updateTriggers function.
+func (p *kafkaPubSub) topicPoller(client *kgo.Client) error {
+	p.closeWg.Add(1)
+	defer p.closeWg.Done()
+
+	for {
+		select {
+		// Close the client if the context was canceled
+		// This is not the context of the subscription, but the context of application shutdown
+		case <-p.ctx.Done():
+			client.Close()
+			return nil
+
+		default:
+			// Try to fetch max records from any subscribed topics
+			// In the future, we could create a client per topic to fetch in parallel
+			fetches := client.PollRecords(p.ctx, 10_000)
+			if fetches.IsClientClosed() {
+				return errClientClosed
+			}
+
+			if errs := fetches.Errors(); len(errs) > 0 {
+
+				for _, fetchError := range errs {
+
+					// If the context was canceled, the error is wrapped in a fetch error
+					if errors.Is(fetchError.Err, context.Canceled) {
+						return nil
+					}
+
+					var kErr *kerr.Error
+					if errors.As(fetchError.Err, &kErr) {
+						if !kErr.Retriable {
+							p.logger.Error("unrecoverable fetch error",
+								zap.Error(fetchError.Err),
+								zap.String("topic", fetchError.Topic),
+							)
+
+							// If the error is not recoverable, return it and abort the poller
+							return fetchError.Err
+						}
+					} else {
+						p.logger.Error("fetch error", zap.Error(fetchError.Err), zap.String("topic", fetchError.Topic))
+					}
+				}
+			}
+
+			iter := fetches.RecordIter()
+			for !iter.Done() {
+				r := iter.Next()
+
+				rec := &record{
+					topic: r.Topic,
+					data:  r.Value,
+				}
+
+				p.updateTriggers(rec)
+			}
+		}
+	}
+}
+
+// updateTriggers updates all subscriptions that are interested in the topic with the new data.
+func (p *kafkaPubSub) updateTriggers(rec *record) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	// Notify all subscriptions that are interested in the topic
+	if sub, ok := p.subscriptions[rec.topic]; ok {
+
+		for _, s := range sub {
+			// If the subscription is still active, update it
+			if s.context.Err() == nil {
+				p.logger.Debug("subscription update", zap.String("topic", rec.topic), zap.ByteString("data", rec.data))
+				s.updater.Update(rec.data)
+			}
+		}
+	}
+}
+
+// Subscribe subscribes to the given topics and updates the subscription updater.
+// The engine already deduplicates subscriptions with the same topics, stream configuration, extensions, headers, etc.
+func (p *kafkaPubSub) Subscribe(ctx context.Context, event pubsub_datasource.KafkaSubscriptionEventConfiguration, updater resolve.SubscriptionUpdater) error {
+
+	p.logger.Debug("subscribe",
+		zap.Strings("topics", event.Topics),
+		zap.String("providerID", event.ProviderID),
+	)
+
+	sub := &subscription{
```

Comment:
I think the concept of subscription here makes the whole thing more complex than it needs to be.

As an alternative, you could create a client in Subscribe, then start the poller, and update through the updater until the ctx is done.

You're optimizing for the case that one Trigger depends on multiple topics, which I think is very rare.